src/ceph-disk

```diff
@@ -2461,6 +2678,316 @@ def main_activate(args):
+                # that because in some situation we can not update
+                # partition table immediately
+                dmcrypt = False
+                part_type = get_partition_type(item)
+                if part_type == DMCRYPT_OSD_UUID or \
+                   part_type == DMCRYPT_LUKS_OSD_UUID:
+                    dmcrypt = True
+                    part_uuid_journal = ''
+                    part_uuid = get_partition_uuid(item)
+                    dev_path = '/dev/mapper/' + part_uuid
+                    whoami, dm_journal = get_osd_id_journal_dm(dev_path, dmcrypt)
+                    part_uuid_journal = get_partition_uuid(dm_journal)
+                elif part_type == OSD_UUID or \
+                     part_type == MPATH_OSD_UUID:
+                    whoami, dm_journal = get_osd_id_journal_dm(item, dmcrypt)
+                else:
+                    continue
+            except:
+                # other cases will return MountError
+                raise MountError
+            # break when we found the target osd_id.
+            # that can avoid handle the redundant checking
+            if whoami == osd_id:
+                found = True
+                (base_dev, part_num) = split_dev_base_partnum(item)
+                break
+        if not found:
+            raise Error('Could not find the partition of osd.%s!' % osd_id)
+    else:
+       if not os.path.exists(args.path):
+           raise Error('%s does not exist' % args.path)
+       else:
+           # check dmcrypt first.
+           part_type = get_partition_type(args.path)
+           if part_type == DMCRYPT_OSD_UUID or \
+              part_type == DMCRYPT_LUKS_OSD_UUID:
+              dmcrypt = True
+              part_uuid_journal = ''
+              part_uuid = get_partition_uuid(args.path)
+              dev_path = '/dev/mapper/' + part_uuid
+           else:
+               # other cases will return args.path
+               dev_path = args.path
+           # mount point is removed, try to mount to tmp.folder
+           osd_id, dm_journal = get_osd_id_journal_dm(dev_path, dmcrypt)
+           if dm_journal:
+               part_uuid_journal = get_partition_uuid(dm_journal)
+           base_dev = get_partition_base(args.path)
+           mount_info.append(args.path)
+
+
+    # Before osd deactivate, we cannot destroy it
+    status_code = _check_osd_status(args.cluster, osd_id)
+    if status_code != OSD_STATUS_OUT_DOWN and \
+       status_code != OSD_STATUS_IN_DOWN:
+        raise Error("Could not destroy the active osd. (osd-id: %s)" % \
+                    osd_id)
+
+    # Remove OSD from crush map
+    _remove_from_crush_map(args.cluster, osd_id)
+
+    # Remove OSD cephx key
+    _delete_osd_auth_key(args.cluster, osd_id)
+
+    # Deallocate OSD ID
+    _deallocate_osd_id(args.cluster, osd_id)
+
+    # we remove the crypt map and device mapper (if dmcrypt is True)
+    if dmcrypt:
```

Comment:
hi @dachary,

About remove the crypt map on the destroy stage.
I prefer to do this operation on here for two reasons:

1. deactivate can be resume to activate, so that could make activate stage do not setup dmcrypt again.
2. if we remove the map on deactivate stage, we also need to setup the dmcrypt again when destroy stage.
    (we need to mount to get some info and confirm the osd to remove)

so I remove the crypt map on the destroy stage to avoid the redundant dmcrypt setup.
would you give me some suggestions abut when we remove the crypt map

thanks!!!
vicente